File: app/src/main/java/com/stupidtree/hichat/ui/welcome/login/LoginFragment.java
package com.stupidtree.hichat.ui.welcome.login;

import androidx.lifecycle.Observer;

import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.stupidtree.hichat.ui.base.BaseFragment;
import com.stupidtree.hichat.R;

import butterknife.BindView;

/**
 * 登录页面Fragment
 */
public class LoginFragment extends BaseFragment<LoginViewModel> {

    /**
     * View绑定区
     */
    @BindView(R.id.username)
    EditText usernameEditText;
    @BindView(R.id.password)
    EditText passwordEditText;
    @BindView(R.id.login)
    Button loginButton;
    @BindView(R.id.loading)
    ProgressBar loadingProgressBar;

    public LoginFragment() {
    }

    public static LoginFragment newInstance(){
        return new LoginFragment();
    }


    @Override
    protected Class<LoginViewModel> getViewModelClass() {
        return LoginViewModel.class;
    }

    @Override
    protected void initViews(View view) {
        //登录表单的数据变更监听器
        viewModel.getLoginFormState().observe(this, loginFormState -> {
            if (loginFormState == null) {
                return;
            }
            //将表单合法性同步到登录按钮可用性
            loginButton.setEnabled(loginFormState.isDataValid());
            //若有表单上的错误，则通知View显示错误
            if (loginFormState.getUsernameError() != null) {
                usernameEditText.setError(getString(loginFormState.getUsernameError()));
            }
            if (loginFormState.getPasswordError() != null) {
                passwordEditText.setError(getString(loginFormState.getPasswordError()));
            }
        });

        //登录结果的数据变更监听
        viewModel.getLoginResult().observe(this, new Observer<LoginResult>() {
            @Override
            public void onChanged(LoginResult loginResult) {
                loadingProgressBar.setVisibility(View.INVISIBLE);
                Toast.makeText(getContext(),loginResult.getMessage(),Toast.LENGTH_SHORT).show();
                if(loginResult.getState()== LoginResult.STATES.SUCCESS){
                    requireActivity().finish();
                }else if(loginResult.getState()== LoginResult.STATES.WRONG_USERNAME) {
                    usernameEditText.setError(getString(loginResult.getMessage()));
                }else if(loginResult.getState()== LoginResult.STATES.WRONG_PASSWORD){
                    passwordEditText.setError(getString(loginResult.getMessage()));
                }
            }
        });

        // 登录表单的文本监视器
        TextWatcher afterTextChangedListener = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // ignore
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // ignore
            }

            @Override
            public void afterTextChanged(Editable s) {
                //将文本信息改变通知给ViewModel
                viewModel.loginDataChanged(usernameEditText.getText().toString(),
                        passwordEditText.getText().toString());
            }
        };
        usernameEditText.addTextChangedListener(afterTextChangedListener);
        passwordEditText.addTextChangedListener(afterTextChangedListener);

        //使得手机输入法上”完成“按钮映射到登录动作
        passwordEditText.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                viewModel.login(usernameEditText.getText().toString(),
                        passwordEditText.getText().toString());
            }
            return false;
        });

        loginButton.setOnClickListener(v -> {
            loadingProgressBar.setVisibility(View.VISIBLE);
            viewModel.login(usernameEditText.getText().toString(),
                    passwordEditText.getText().toString());
        });
    }


    @Override
    protected int getLayoutId() {
        return R.layout.fragment_login;
    }
}

File: app/src/main/java/com/stupidtree/cloudliter/ui/welcome/login/LoginFragment.kt
package com.stupidtree.cloudliter.ui.welcome.login

import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.*
import androidx.lifecycle.Observer
import butterknife.BindView
import com.stupidtree.cloudliter.R
import com.stupidtree.cloudliter.ui.base.BaseFragment

/**
 * 登录页面Fragment
 */
class LoginFragment : BaseFragment<LoginViewModel>() {
    /**
     * View绑定区
     */
    @JvmField
    @BindView(R.id.username)
    var usernameEditText: EditText? = null

    @JvmField
    @BindView(R.id.password)
    var passwordEditText: EditText? = null

    @JvmField
    @BindView(R.id.login)
    var loginButton: Button? = null

    @JvmField
    @BindView(R.id.loading)
    var loadingProgressBar: ProgressBar? = null
    override fun getViewModelClass(): Class<LoginViewModel> {
        return LoginViewModel::class.java
    }

    override fun initViews(view: View) {
        //登录表单的数据变更监听器
        viewModel!!.loginFormState.observe(this, Observer { loginFormState: LoginFormState ->
            //将表单合法性同步到登录按钮可用性
            loginButton!!.isEnabled = loginFormState.isDataValid
            //若有表单上的错误，则通知View显示错误
            if (loginFormState.usernameError != null) {
                usernameEditText!!.error = getString(loginFormState.usernameError!!)
            }
            if (loginFormState.passwordError != null) {
                passwordEditText!!.error = getString(loginFormState.passwordError!!)
            }
        })

        //登录结果的数据变更监听
        viewModel!!.loginResult.observe(this, Observer { loginResult ->
            loadingProgressBar!!.visibility = View.INVISIBLE
            if (loginResult != null) {
                Toast.makeText(context, loginResult.message, Toast.LENGTH_SHORT).show()
            }
            if (loginResult != null) {
                if (loginResult.state == LoginResult.STATES.SUCCESS) {
                    requireActivity().finish()
                } else if (loginResult.state == LoginResult.STATES.WRONG_USERNAME) {
                    usernameEditText!!.error = getString(loginResult.message)
                } else if (loginResult.state == LoginResult.STATES.WRONG_PASSWORD) {
                    passwordEditText!!.error = getString(loginResult.message)
                }
            }
        })

        // 登录表单的文本监视器
        val afterTextChangedListener: TextWatcher = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                // ignore
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                // ignore
            }

            override fun afterTextChanged(s: Editable) {
                //将文本信息改变通知给ViewModel
                viewModel!!.loginDataChanged(usernameEditText!!.text.toString(),
                        passwordEditText!!.text.toString())
            }
        }
        usernameEditText!!.addTextChangedListener(afterTextChangedListener)
        passwordEditText!!.addTextChangedListener(afterTextChangedListener)

        //使得手机输入法上”完成“按钮映射到登录动作
        passwordEditText!!.setOnEditorActionListener { v: TextView?, actionId: Int, event: KeyEvent? ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                viewModel!!.login(usernameEditText!!.text.toString(),
                        passwordEditText!!.text.toString())
            }
            false
        }
        loginButton!!.setOnClickListener { v: View? ->
            loadingProgressBar!!.visibility = View.VISIBLE
            viewModel!!.login(usernameEditText!!.text.toString(),
                    passwordEditText!!.text.toString())
        }
    }

    override fun getLayoutId(): Int {
        return R.layout.fragment_login
    }

    companion object {
        fun newInstance(): LoginFragment {
            return LoginFragment()
        }
    }
}